### PR TITLE
Create display conditionals for the Feedback Request Card

### LINKED ISSUE #2660: [IAFeedback] Implement feedback request display logic
To make it much easier to test the rest of the functionality being added for this project, I've broken out this functionality to be performed last after everything else is implemented.

**Note**: Ticket #2645  should be implemented prior to starting this task.

## When To Show It

> We’d like users to have had time to interact with the app before providing feedback, and we also don’t want to annoy them too much by polling them too often.

1. Users who’ve installed the app at least 3 months ago
2. Once every 6 months

## When It Disappears

Pressing any of the buttons will hide the section.

### PR DESCRIPTION
**Pull request by ThomazFB**
Summary
==========
Fixes issue #2660 by making available a set of rules and conditionals with the following rules:

- The known installation date is more than 3 months ago
- The user has not given feedback for more than 6 months ago.

Changelog
==========
| Description | Commit |
| ------------- | ------------- |
| Create a extension property for Date objects where's possible to determine the amount of days from that Date object compared to today Date | b688474d14a1ebe570d038cb803337ecd21e9061 |
| Create SharedPreferences configuration to store the epoch timestamp as String within the `AppPrefs` object | 7241975deeae5056ff4d4381c098a68b90cb1050 |
| Create an infix function to work as a `greaterThan` operator that's capable to operate with nullable values | c895c280d1d3c428697c13b57cb4a53cb441cc10 |
| Add the conditionals to verify if the app should request feedback from the user | b69330fd395632ee6580ae01a757352bb8177262 |
| Insert the Feedback Due conditionals to the FeedbackRequestCard calls inside the `MyStoreFragment` and `DashboardFragment` | 5af895f236e864b418961636a868ad6ac17ba9f2 |
| Refactor the `MyStoreFragment` and `DashboardFragment` to better handle the visibility of the card whenever the device Local Date time is changed and modified | ad9a658cef3ea4b90810c2159043dff0e25092f2 |

Screenshots
==========
| Feedback Request Card Hidden  | Feedback Request Card Visible |
| ------------- | ------------- |
| screencap-2020-09-02T065522 019Z | screencap-2020-09-02T065603 035Z |

How to Test
==========
Scenario 1 - App installed for more than three months ago and no feedback was given
------------
1. Go to the device settings and change the system calendar to more than 90 days *BEFORE* of your local current date.
2. Make a fresh install of the Woo app from the scratch
3. Go back to the device settings and reactivate the automatic system calendar definition back to your local current date.
4. Go through the Login flow and verify if the Feedback card is visible
5. Verify if the card gets dismissed after any of both actions gets clicked
6. Verify if the card doesn't come back after reopening the app or through any kind of app interaction

Scenario 2 - App freshly installed and no feedback was given
------------
1. Make a fresh install of the Woo app from the scratch
2. Go through the Login flow and verify if the Feedback card is *NOT* visible
3. Verify if the card doesn't show up after reopening the app or through any kind of app interaction

Scenario 3 - Last Feedback was given six months ago
------------
1. Run through Scenario 1
2. Go to the device settings and change the system calendar to more than 180 days *AFTER* of your local current date.
3. Go back to the Woo app
4. Verify if the card is visible and clickable
5. Verify if the card gets dismissed after any of both actions gets clicked
6. Verify if the card doesn't come back after reopening the app or through any kind of app interaction

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt [hunk 1]
```diff
@@ -63,13 +65,37 @@ object AppPrefs {
...
     }
 
+    var installationDate: Date?
+        get() = getString(UndeletablePrefKey.APP_INSTALATION_DATE)
+            .toLongOrNull()
+            ?.let { Date(it) }
+
+        private set(value) = value
+            ?.time.toString()
+            .let { setString(UndeletablePrefKey.APP_INSTALATION_DATE, it) }
```
Comment (shiki) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt]: @ThomazFB In iOS, we aimed to start gathering feedback right away from users who have the app on their devices for more than 3 months already. 

To achieve this, [we used one of the app's directories on the device](https://github.com/woocommerce/woocommerce-ios/blob/609f5efa0e6fca479fbffb1abd486a9846816a59/Yosemite/Yosemite/Stores/AppSettings/InAppFeedbackCardVisibilityUseCase.swift#L105-L117) to infer the _true_ installation date. If that didn't work, we used the `installationDate` setting as a backup. 

Do you think this is something we should do here too? It looks like there's also [a hacky way](https://stackoverflow.com/questions/5311644/get-application-installed-date-on-android) to get the installation date for Android.

**Review by shiki — COMMENTED** [review]
Just dropping by the leave a question. 🙂

**Comment by shiki** [comment]
@ThomazFB When using this branch, I also noticed that the constraints described in https://github.com/woocommerce/woocommerce-android/issues/2641 are not implemented yet. 

For the I Like It button:

1. Tapping on Later → the user will keep seeing the “Enjoying the WC app” card. Currently, the card disappears right away. 
2. No Thanks → the card will be dismissed. ✔️ 

This is going to be done in a separate PR, right?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt (hunk 1 above)
Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt]: Hey @shiki! Really awesome point, thanks for pointing that out, I've tested the consistency of the installation date acquired through the `PackageManager` and looks reliable and functional for what we're trying to achieve, in that case, I've changed the way we define the installation date right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/51fefb82df1f2759463d63c9d014b1d20d3dd9bb, let me know what you think.

**Comment by ThomazFB** [comment]
> When using this branch, I also noticed that the constraints described in #2641 are not implemented yet.
>
> For the I Like It button:
>
> Tapping on Later → the user will keep seeing the “Enjoying the WC app” card. Currently, the card disappears right away.
> No Thanks → the card will be dismissed. ✔️
> This is going to be done in a separate PR, right?

Hey @shiki , thanks a lot for your feedback. Actually I think it makes sense to adjust this problem right away on this PR. Since this problem was introduced here and this PR changes how the `MyStoreFragment` and `DashboardFragment` handles when the feedback was given or not. I've adjusted this point right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/1369d09c248e76d3f22c0a9bbed247dae87c75b5. Let me know what you think.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt (hunk 1 above)
Comment (shiki) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt]: Looks good to me! 🤸

**Review by shiki — COMMENTED** [review]
It's going nicely. The card is no longer dismissed when dismissing the App Rating dialog by touching outside of it. I have noticed one thing:

### The Later button dismisses the card. 

I performed these steps:

1. Reinstalled the app. 
2. Time-traveled to 3 months. 
3. Tapped on I Like It. 
4. Tapped on Later. 

The card was dismissed but it shouldn't have been. Only the No Thanks and Rate Now button should dismiss the card. 🙂

**Comment by ThomazFB** [comment]
>It's going nicely. The card is no longer dismissed when dismissing the App Rating dialog by touching outside of it. I have >noticed one thing:
>
>The Later button dismisses the card.
>
>I performed these steps:
>
>Reinstalled the app.
>Time-traveled to 3 months.
>Tapped on I Like It.
>Tapped on Later.
>The card was dismissed but it shouldn't have been. Only the No Thanks and Rate Now button should dismiss the card. 🙂

Hey @shiki, thanks for taking some time to review it again! For that scenario that behavior was intentional, wouldn't be desirable that if the user selects `later` we will only prompt the banner again the next time he opens up the app? I thought if the user decides to select `later` and we still keep showing up the banner it will look like we're insisting on asking the review 😅 WDYT? @Garance91540 have some take on that use case?

**Comment by Garance91540** [comment]
> Hey @shiki, thanks for taking some time to review it again! For that scenario that behavior was intentional, wouldn't be desirable that if the user selects later we will only prompt the banner again the next time he opens up the app? I thought if the user decides to select later and we still keep showing up the banner it will look like we're insisting on asking the review 😅 WDYT? @Garance91540 have some take on that use case?

Yeah my main concern with that new feature is that some users might get annoyed by it and complain about it. So if users select "Later" here,  I think it's ok to play it safe and display it again the next time they open up the app

**Comment by shiki** [comment]
Sounds good to me! Thanks for confirming, Garance.

**Comment by nbradbury** [comment]
@ThomazFB I'm not sure if this problem is caused by this branch, but when I select "Could be better" then dismiss the feedback WebView, I'm left with a blank dashboard. I reproduced this on two different emulators.

ezgif com-video-to-gif

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt [hunk 2]
```diff
@@ -63,13 +65,61 @@ object AppPrefs {
         // The app update for this version was cancelled by the user
         CANCELLED_APP_VERSION_CODE,
         // Application permissions
-        ASKED_PERMISSION_CAMERA
+        ASKED_PERMISSION_CAMERA,
+        // Date of the app installation
+        APP_INSTALATION_DATE,
+        // Date of the last time the user sent feedback on the app
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt]: Is this really the date the user last sent feedback? Looking at the code it seems like it's the date the user was last _asked_ for feedback.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt [hunk 3]
```diff
@@ -33,6 +36,13 @@ object DateUtils {
         SimpleDateFormat("yyyy-MM-dd", Locale.US)
     }
 
+    private const val THREE_MONTHS_IN_DAYS = 90
+    private const val SIX_MONTHS_IN_DAYS = 180
+
+    val userFeedbackIsDue: Boolean
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt]: This is a nitpick,  but `DateUtils` seems an odd place for this. Maybe`AppPrefs` instead?

**Comment by nbradbury** [comment]
@ThomazFB I'm finished with my review, leaving just a few minor comments. I'll move on to #2822 so we can get that ready to merge as well.

**Comment by loremattei** [comment]
Hi there! I'm moving this to 5.1 because 5.0 has been cut. If you want this to make it to 5.0, please feel free to ping me.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt (hunk 3 above)
Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt]: That's a really good point, thanks for that, I wasn't sure of which place should be the better one to leave that conditional. I followed your suggestion and moved to the `AppPrefs` right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/74fb443270ef3288784c4988afe16488e8ca32e2. But one noteworthy point is that at the upcoming https://github.com/woocommerce/woocommerce-android/pull/2822 PR I've refactored all the feedback preferences to a new `FeedbackPrefs` object, when this change right here is synced there, I think that would be the best option to leave that kind of logic. WDYT?

File: WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt (hunk 2 above)
Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefs.kt]: That's true, the current naming isn't describing correctly the represented information, I've changed this issue right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/e9ab333086c6094851733a10ed0ac471c17b731b. Thanks for your feedback, Nick!

**Comment by ThomazFB** [comment]
>  I'm not sure if this problem is caused by this branch, but when I select "Could be better" then dismiss the feedback WebView, I'm left with a blank dashboard. I reproduced this on two different emulators.

Hey @nbradbury ! Thank you so much for your review! Really nice catch, from what I've discovered, it looks like the `MainActivity` wasn't ready to deal with the navigation controller bringing the UI back to the `My Store` section, so I've fixed that by always making sure that the active top-level Fragment must be visible once a navigation destination is set back to the root. I've made this change right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/446c2e14b03782b114e55266a5eb45cf9b88e2ab. Let me know how this goes.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt (hunk 3 above)
Comment (nbradbury) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt]: Sounds good to me!

**Comment by nbradbury** [comment]
> I've fixed that by always making sure that the active top-level Fragment must be visible once a navigation destination is set back to the root.

@ThomazFB This definitely works, but I'm left wondering why the top-level fragment is invisible in the first place. That doesn't happen with any other child fragment. Any ideas?

**Comment by ThomazFB** [comment]
Hey @nbradbury, nevermind my previous solution, I kept digging on what was going on to make the Top Level Fragment become hidden and your guess is correct, this was a bug introduced by this PR right here, I've incorrectly referenced the Feedback Card visibility and caused the whole fragment to become hidden, it really took me a long while to detect this simple mistake, but it's fixed now, thank you for your awesome catch on that! I've submitted the fix right here: https://github.com/woocommerce/woocommerce-android/pull/2812/commits/5db737871fff7c2bc1b0cd787c08d25f22506d36

**Review by nbradbury — APPROVED** [review]
Looks good, nice work @ThomazFB! :shipit:

**Pull request merged**